### ISSUE TITLE
Enable support for image launching on macOS and Windows

### ISSUE THREAD

**Issue by moondev** [canonical/multipass#1260]
Is it possible to directly launch a qcow2/img cloud-init enabled image on macOS? This works great on Linux. Is this due to differences between qemu and hyperkit image format?

Test image used: https://download.fedoraproject.org/pub/fedora/linux/releases/31/Cloud/x86_64/images/Fedora-Cloud-Base-31-1.9.x86_64.qcow2

works on linux
```
multipass launch -n fedora https://download.fedoraproject.org/pub/fedora/linux/releases/31/Cloud/x86_64/images/Fedora-Cloud-Base-31-1.9.x86_64.qcow2
```

fails on macos
```
returns "http and file based images are not supported"
```

**Comment by Saviq** [comment]
Hi @moondev yeah, we currently have to pass kernel + initrd directly to hyperkit (no bootloader), so we can't really support images that we can't be sure will work with the kernels we use…

We're working on it, though!

**Comment by moondev** [comment]
@Saviq Thanks for the details! Very interesting as well as clever way to leverage hyperkit for cloud-init.

Looking forward to a future update to support this. I suppose using an alternative hypervisor could be an option but then you would need to install it first which is no fun. Congrats on the `1.0.0` launch and keep up the great work!

**Comment by alexellis** [comment]
Would qemu work as an alternative with its hardware acceleration?

**Comment by Saviq** [comment]
Hi @alexellis we'll be evaluating `qemu` for that reason in the near future.

**Comment by alexellis** [comment]
How are you planning on dealing with networking? I had issues getting any examples for qemu on MacOS to work for port mapping.

**Comment by jasonmccallister** [comment]
Ideally I would love to be able to build a custom prebuilt ubuntu image and download them from the launch command. We have a hefty cloud-config setup right now that would reduce our setup by just downloading an image instead.

**Comment by madalinignisca** [comment]
Wouldn't it be easy to add a parameter flag to enable "experiment" mode and allow it in Windows/MacOS, and we users would be responsible if the image would work or not? I don't understand at all the reason to forbid Windows and MacOS users to have this feature even if they explicitly would accept it as experimental and unsupported.

To give an example: This blocks me right now to do a large demo of something that works really nice on my personal laptop with Ubuntu as main host, but can't run it on my work laptop with Windows, and my company forbids to use another device for the demo purely of some company policies.
Are you considering that with this kind of things you are blocking us developers and engineers to convince our companies for (more) adoption of Ubuntu and Canonical products and services?

**Comment by mkhon** [comment]
I succeeded substituting Ubuntu disk image with CentOS 7 cloud image on macOS:

1) create Ubuntu 20 multipass VM ("multipass launch ..."), then stop it

`multipass launch -n <name>
 multipass stop <name>`

2) download CentOS cloud image (I used this one: https://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-2009.qcow2)

3) copy the image to instance vault:

`cp CentOS-7-x86_64-GenericCloud-2009.qcow2 /var/root/Library/Application Support/multipassd/vault/instances/<name>`

4) delete existing disk image and make a symlink:

`rm ubuntu-20.04-server-cloudimg-amd64.img ubuntu-20.04-server-cloudimg-amd64.img
ln -s CentOS-7-x86_64-GenericCloud-2009.qcow2 ubuntu-20.04-server-cloudimg-amd64.img`

The contents of the instance vault directory should look like:

`-rw-r--r--  1 root  wheel  931856384 Jun 10 13:31 CentOS-7-x86_64-GenericCloud-2009.qcow2
-rw-r--r--  1 root  wheel      53248 Jun 10 13:15 cloud-init-config.iso
-rw-r--r--  1 root  wheel         12 Jun 10 13:24 pty
-rw-r--r--  1 root  wheel   27621125 Jun  8 15:50 ubuntu-20.04-server-cloudimg-amd64-initrd-generic
-rw-r--r--  1 root  wheel   13668608 Jun  8 15:50 ubuntu-20.04-server-cloudimg-amd64-vmlinuz-generic
lrwxr-xr-x  1 root  wheel         39 Jun 10 13:23 ubuntu-20.04-server-cloudimg-amd64.img -> CentOS-7-x86_64-GenericCloud-2009.qcow2`

5) start the instance

`multipass start <name>`

The instance uses Ubuntu 20 kernel and initrd which should be suitable for almost any recent Linux cloud image

**Comment by townsend2010** [comment]
Hey @mkhon,

That's quite enterprising to say the least:wink:

**Comment by goofrider** [comment]
Now that Mac OS version of Multipass is switching to Qemu, would the team consider enabling instance creation from custom URLs/ISOs?

**Comment by BurlyLuo** [comment]
+1

**Comment by BurlyLuo** [comment]
Do we have a plan to support CentOS?

**Comment by atomlab** [comment]
Hello! Where do I need to place the downloaded Debian Cloud image to run it on MacOS with M1 processor?
https://cloud.debian.org/images/cloud/bookworm/daily/latest/debian-12-genericcloud-arm64-daily.qcow2

**Comment by underrun** [comment]
> > Since this runs on QEMU and not hyperkit, any chance we could get image launching support for the Apple Silicon release?
> > #1260
> 
> _Originally posted by @m-bers in [#1857 (comment)](https://github.com/canonical/multipass/issues/1857#issuecomment-950114617)_
> 
> @m-bers, all, it would be near impossible for us to ensure the user experience with custom images. There are certain requirements of the image to be compatible with Multipass. Those of you requesting it would probably be fine with it, but others may be surprised and frustrated that "random" images downloaded from the Internet don't work.

@Saviq I don't think this argument holds up while multipass supports launching images from files on linux. If I download some random image from the internet and try to start it on linux, canonical has exactly as much control over the user experience in that situation as it does on macos or windows.

The users of multipass on mac/win are likely mostly developers. The users that even want to try building a qcow2 image out of an iso for some non ubuntu cloud image probably know what they are doing enough to be aware of the pitfalls and debug them. And it's pretty likely those users are trying to do this to make their lives easier via the really nice ux multipass provides.

In other words, by not supporting file based images on mac/win you are ensuring users will have a much worse experience working with their custom images.

Please address this issue.

**Comment by madalinignisca** [comment]
Thank you!

> Since I have been following this thread for over 2 years, I found an alternative solution to run other distro's in the same fashion as Multipass does, with [Lima](https://lima-vm.io/).
> 
> 
> 
> After [installing Lima](https://lima-vm.io/docs/installation/), you can create a Debian VM as such: 
> 
> ```limactl create --name debian-12 https://raw.githubusercontent.com/lima-vm/lima/master/examples/debian-12.yaml```
> 
> or use `https://github.com/lima-vm/lima/blob/master/examples/ubuntu-lts.yaml` to run Ubuntu.
> 
> 
> 
> Start your VM with `limactl start debian-12`.
> 
> 
> 
> And then shell into your vm with `limactl shell debian-12`.
> 
> 
> 
> 
> 
> @canonical, if I violate the terms of service, please just remove my comment.

**Comment by emakarov** [comment]
currently i was exploring the flow of creating a packer image from our production docker image and then launch it with multipass on Mac. and also stucked with the same limitation. While workaround exists, it looks like a very strange limitation, especially if workaround really works (in progress of checking this)
In regards to things like Lima, @OhMyMndy by any chance if you know, does it work with rosetta mode on Mac M1/M2?

**Comment by ibo-g** [comment]
What about doing work arounds with older official images provided by Canonical, which you were including to instances and allowing to use in multipass back then? Do you think that can we use them properly without problem in theory, after applying work arounds, since once upon a time in past, multipass was already running it?

**Comment by animeshdas** [comment]
> I succeeded substituting Ubuntu 20 disk image with CentOS 7 cloud image on macOS:
> 
> 1. create Ubuntu 20 multipass VM, then stop it
> 
> ```
> multipass launch -n <name>
> multipass stop <name>
> ```
> 
> 2. download CentOS cloud image (I used this one: https://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-2009.qcow2c)
> 3. convert the image and store into instance vault:
> 
> ```
> "/Library/Application Support/com.canonical.multipass/bin/qemu-img-hyperkit" convert -O qcow2 CentOS-7-x86_64-GenericCloud-2009.qcow2c "/var/root/Library/Application Support/multipassd/vault/instances/<name>/CentOS-7-x86_64-GenericCloud-2009.qcow2"
> ```
> 
> 4. move out existing disk image and make a symlink:
> 
> ```
> mv ubuntu-20.04-server-cloudimg-amd64.img ubuntu-20.04-server-cloudimg-amd64.img~
> ln -s CentOS-7-x86_64-GenericCloud-2009.qcow2 ubuntu-20.04-server-cloudimg-amd64.img
> ```
> 
> The contents of the instance vault directory should look like:
> 
> ```
> -rw-r--r--  1 root  wheel  931856384 Jun 10 13:31 CentOS-7-x86_64-GenericCloud-2009.qcow2
> -rw-r--r--  1 root  wheel      53248 Jun 10 13:15 cloud-init-config.iso
> -rw-r--r--  1 root  wheel         12 Jun 10 13:24 pty
> -rw-r--r--  1 root  wheel   27621125 Jun  8 15:50 ubuntu-20.04-server-cloudimg-amd64-initrd-generic
> -rw-r--r--  1 root  wheel   13668608 Jun  8 15:50 ubuntu-20.04-server-cloudimg-amd64-vmlinuz-generic
> lrwxr-xr-x  1 root  wheel         39 Jun 10 13:23 ubuntu-20.04-server-cloudimg-amd64.img -> CentOS-7-x86_64-GenericCloud-2009.qcow2
> ```
> 
> 5. start the instance
> 
> ```
> multipass start <name>
> ```
> 
> The instance uses Ubuntu 20 kernel and initrd which should be suitable for almost any recent Linux cloud image
> 
> Additionally, you may need to copy out /lib/modules from Ubuntu image (e.g. boot Ubuntu instance, copy /lib/modules out, then boot your image, copy /lib/modules back) so that kernel modules in the image match the kernel used by multipass.
> 
> Make sure you umount /lib/modules (or /usr/lib/modules) before copying it back because it is mounted (I guess by initird) as tmpfs (copymods):
> 
> ```
> # umount -l /usr/lib/modules
> # systemctl restart systemd-udevd
> ```

This works just to boot the image. But networking, file sharing aren't working. I tried with AlmaLinux 9.6 qcow2 cloud generic image, it booted but networking, sharing and other stuff not working. Upon replacing the /lib/modules it failed to boot.

**Comment by ricab** [comment]
Hi all. This is now done and about to be released. You can try it out with the [1.16.0 RC](https://github.com/canonical/multipass/releases/tag/v1.16.0-rc3).

**Comment by mkhon** [comment]
> This works just to boot the image. But networking, file sharing aren't working. I tried with AlmaLinux 9.6 qcow2 cloud generic image, it booted but networking, sharing and other stuff not working. Upon replacing the /lib/modules it failed to boot.

This is likely due to AlmaLinux 9.6 linux kernel (5.14?) is newer than Ubuntu 20 linux kernel (5.4?) and likely AlmaLinux 9.6 userland expecting some 5.14 kernel features which do not exist in Ubuntu 20 kernel.

You can try to use Ubuntu 22 as a base and see if it helps.

**Comment by animeshdas** [comment]
> > This works just to boot the image. But networking, file sharing aren't working. I tried with AlmaLinux 9.6 qcow2 cloud generic image, it booted but networking, sharing and other stuff not working. Upon replacing the /lib/modules it failed to boot.
> 
> This is likely due to AlmaLinux 9.6 linux kernel (5.14?) is newer than Ubuntu 20 linux kernel (5.4?) and likely AlmaLinux 9.6 userland expecting some 5.14 kernel features which do not exist in Ubuntu 20 kernel.
> 
> You can try to use Ubuntu 22 as a base and see if it helps.

To be honest, that not a solution everyone is discussing about here. I know I can run Ubuntu, and that's quite obvious. 

My point is; if we can run some other distro image qcow2 cloud-init enabled image or not? The tool should be vendor neutral else what is the point of using open source software, if it restricts or dictate my choice of OS. 

For universal adoption and meaningful ecosystem, it should support multi-vendor OS, not only just ubuntu. At least few major other distro and their derivatives.  Else mass adoption would be a real challange. 

REMEMBER not everyone uses Ubuntu for all their use cases.  :-)

**Comment by mkhon** [comment]
> To be honest, that not a solution everyone is discussing about here. I know I can run Ubuntu, and that's quite obvious.
> 
> My point is; if we can run some other distro image qcow2 cloud-init enabled image or not? The tool should be vendor neutral else what is the point of using open source software, if it restricts or dictate my choice of OS.
> 
> For universal adoption and meaningful ecosystem, it should support multi-vendor OS, not only just ubuntu. At least few major other distro and their derivatives. Else mass adoption would be a real challange.
> 
> REMEMBER not everyone uses Ubuntu for all their use cases. :-)

1) I do not represent Canonical

2) The hack I proposed was to use existing Ubuntu 20 VM (kernel) with another cloud-init image with Linux distro of your choice (CentOS in my example)

3) My suggestion about your AlmaLinux experience (kernel boots but no networking) was to try to use Ubuntu 22 VM as a base (in my example I used Ubuntu 20 VM as a base), which has the kernel newer than your AlmaLinux 9.6 (Ubuntu 20 has older kernel)

**Comment by felipecrs** [comment]
> My point is; if we can run some other distro image qcow2 cloud-init enabled image or not? The tool should be vendor neutral else what is the point of using open source software, if it restricts or dictate my choice of OS.
> 
> For universal adoption and meaningful ecosystem, it should support multi-vendor OS, not only just ubuntu. At least few major other distro and their derivatives. Else mass adoption would be a real challange.
> 
> REMEMBER not everyone uses Ubuntu for all their use cases. :-)

Obviously, it would be great if Canonical supported a thousand different distributions in Multipass, but I think it's fair if they have chosen not to.

Anyone is free to fork Multipass and add support for whatever other distro they want. That's effectively open source.

PS: I also do not represent Canonical.

**Comment by holta** [comment]
`multipass launch https://cloud.debian.org/images/cloud/trixie/daily/latest/debian-13-generic-amd64-daily.qcow2` appears to work fine — which is really great news — Multipass 1.16 Release Candidates are a **dramatic improvement** on Windows and Mac!

**Comment by animeshdas** [comment]
upgraded to multipass 1.16.
`multipass launch \                                                                ─╯
  --name almalinux9 \
  --cloud-init ./AlmaLinux9-cloud-init.yaml \
  --disk 5G \
  --memory 2G \
  --cpus 2 \
  file://AlmaLinux-9-GenericCloud-latest.aarch64.qcow2
launch failed: Requested disk (5368709120 bytes) below minimum for this image (10737418240 bytes)`

it failed.
